contrib/babelfishpg_tsql/src/hooks.c

```diff
@@ -556,11 +580,226 @@ pltsql_GetNewObjectId(VariableCache variableCache)
 
 	minOid = atooid(babelfish_dump_restore_min_oid);
 	Assert(OidIsValid(minOid));
-	if (ShmemVariableCache->nextOid >= minOid + 1)
+	if (variableCache->nextOid >= minOid + 1)
 		return;
 
-	ShmemVariableCache->nextOid = minOid + 1;
-	ShmemVariableCache->oidCount = 0;
+	variableCache->nextOid = minOid + 1;
+	variableCache->oidCount = 0;
+}
+
+static Oid
+pltsql_GetNewTempObjectId()
+{
+	Oid			result;
+	Oid			tempOidStart;
+	static Oid 	nextTempOid = InvalidOid;
+
+	/* safety check, we should never get this far in a HS standby */
+	if (RecoveryInProgress())
+		elog(ERROR, "cannot assign OIDs during recovery");
+
+	if (!OidIsValid(BUFFER_START_TO_OID)) /* InvalidOid means it needs assigning */
+	{
+		/* First check to see if another connection has already picked a start, then update. */
+		LWLockAcquire(OidGenLock, LW_EXCLUSIVE);
+		if (OidIsValid(ShmemVariableCache->tempOidStart))
+		{
+			temp_oid_buffer_start = OID_TO_BUFFER_START(ShmemVariableCache->tempOidStart);
+			nextTempOid = ShmemVariableCache->tempOidStart;
+		}
+		else
+		{
+			/* We need to pick a new start for the buffer range. */
+			tempOidStart = ShmemVariableCache->nextOid;
+
+			/*
+			 * Decrement ShmemVariableCache->oidCount to take into account the new buffer we're allocating
+			 */
+			if (ShmemVariableCache->oidCount < temp_oid_buffer_size)
+				ShmemVariableCache->oidCount = 0;
+			else
+				ShmemVariableCache->oidCount -= temp_oid_buffer_size;
+
+			/*
+			 * If ShmemVariableCache->nextOid is below FirstNormalObjectId then we can start at FirstNormalObjectId here and
+			 * GetNewObjectId will return the right value on the next call.  
+			 */
+			if (tempOidStart < FirstNormalObjectId)
```

Comment:
for regular get oid, we set oid count to 0 in case of such wraparound. However, we are not doing the same here

Reply:
Why would we do that here? We have no guarantee that the oidCount should be 0 at this point. That is handled in the above if statement. For instance, if we are at OID wraparound but in fact most of the OID buffer has already been freed, then oidCount will be very large.